---
title: "StarBioTrek:Application Examples"
author: Claudia Cava, Antonio Colaprico, Alex Graudenzi, Gloria Bertoli,Tiago C. Silva,Catharina
  Olsen,Houtan Noushmehr, Gianluca Bontempi, Giancarlo Mauri, Isabella Castiglioni
date: '`r Sys.Date()`'
output: pdf_document
vignette: >
    %\VignetteIndexEntry{SpidermiR_pdf}
    %\VignetteEngine{knitr::rmarkdown}
    \usepackage[utf8]{inputenc}

references:
- author:
  - family: Warde-Farley D, Donaldson S, Comes O, Zuberi K, Badrawi R, and others
    given: null
  id: ref1
  issued:
    year: 2010
  journal: Nucleic Acids Res.
  number: 2
  pages: 214-220
  title: The Gene Mania prediction server biological network integration for gene
    prioritization and predicting gene function
  volume: 38
- author:
  - family: Jiang Q, Wang Y, Hao Y, Juan L, Teng M, Zhang X, Li M, Wang G, Liu Y.
    given: null
  id: ref2
  issued:
    year: 2009
  journal: Nucleic Acids Res.
  number: 1
  pages: 98-104
  title: miR2Disease a manually curated database for microRNA deregulation in human
    disease.
  volume: 37
- author:
  - family: Dweep H, Sticht C, Pandey P, Gretz N.
    given: null
  id: ref3
  issued:
    year: 2011
  journal: Journal of Biomedical Informatics
  number: 1
  pages: 839-7
  title: miRWalk - database prediction of possible miRNA binding sites by "walking"
    the genes of 3 genomes.
  volume: 44
- author:
  - family: Russo F, Di Bella S, Nigita G, Macca V, Lagana A, Giugno R, Pulvirenti
      A, Ferro A.
    given: null
  id: ref4
  issued:
    year: 2012
  journal: PLoS ONE
  number: 10
  pages: e47786
  title: miRandola Extracellular Circulating microRNAs Database.
  volume: 7
- author:
  - family: Csardi G, Nepusz T.
    given: null
  id: ref5
  issued:
    year: 2006
  journal: InterJournal
  number: null
  pages: 1695
  title: The igraph software package for complex network research.
  volume: Complex Systems
- author:
  - family: Rukov J, Wilentzik R, Jaffe I, Vinther J, Shomron N.
    given: null
  id: ref6
  issued:
    year: 2013
  journal: Briefings in Bioinformatics
  number: 4
  pages: 648-59
  title: Pharmaco miR linking microRNAs and drug effects.
  volume: 15

---

# Introduction 

n this vignette, we demonstrate some applications of `SpidermiR` as tool for the study of miRNA network. For basic use of the
`SpidermiR` package, please refer to the vignette `Working with StarBioTrek package`.



# SpidermiR Downstream Analysis: Case Studies
## Case Study n.1: Role of miRNAs in shared protein domains network in Prostate Cancer

In this case study, we downloaded shared protein domains network in Homo Sapiens, using SpidermiRquery, SpidermiRprepare, and SpidermiRdownload.   

Then, we focused on role of miRNAs in this network. We integrated miRNA information using SpidermiRanalyze. We obtained a big network, and in order to understand the underlying biological process of a set of biomarker of interest (e.g. from lab) we performed an analysis to identify their neighbor biomarkers in the shared protein domains network. 

SpidermiRvisualize was used to see the results.

```{r, eval = FALSE}

org<-SpidermiRquery_species(species)
net_shar_prot<-SpidermiRquery_spec_networks(organismID = org[6,],network = "SHpd")
out_net<-SpidermiRdownload_net(net_shar_prot)
geneSymb_net<-SpidermiRprepare_NET(organismID = org[6,],data = out_net)
miRNA_complNET<-SpidermiRanalyze_mirna_gene_complnet(data=geneSymb_net,
                                                      disease="prostate cancer")

biomark_of_interest<-read.delim("C:/Users/UserInLab05/Google Drive/MIRNA AND GENEMANIA/
                                   1 case study/deg_prostate.txt",header=FALSE)
subnet<-SpidermiRanalyze_direct_subnetwork(data=miRNA_complNET,BI=biomark_of_interest$V1)
comm2<-  SpidermiRanalyze_Community_detection(data=subnet,type="FC") 
cd_net<-SpidermiRanalyze_Community_detection_net(data=subnet,comm_det=comm2,size=2)
SpidermiRvisualize_mirnanet(data=cd_net)
miRNA_NET<-SpidermiRanalyze_mirna_network(data=geneSymb_net,disease="prostate cancer")

cd_net_miRNA<-SpidermiRanalyze_Community_detection_net(data=miRNA_NET,comm_det=comm2,size=2)

SpidermiRvisualize_mirnanet(data=cd_net_miRNA)



```

## Case Study n.2: miRNAs regulating degree centrality genes in physical interactions network in breast cancer

```{r, eval = FALSE}
org<-SpidermiRquery_species(species)
net_PHint<-SpidermiRquery_spec_networks(organismID = org[6,],network = "PHint")
out_net<-SpidermiRdownload_net(net_PHint)
geneSymb_net<-SpidermiRprepare_NET(organismID = org[6,],data = out_net)
ds<-do.call("rbind", geneSymb_net)
data1<-as.data.frame(ds[!duplicated(ds), ]) 

sdas<-cbind(data1$gene_symbolA,data1$gene_symbolB)

miRNA_NET<-SpidermiRanalyze_mirna_network(data=geneSymb_net,disease="breast cancer")
topwhol<-SpidermiRanalyze_degree_centrality(sdas)
top10_cent_gene<-SpidermiRanalyze_degree_centrality(miRNA_NET)
miRNA_degree<-top10_cent_gene[grep("hsa",top10_cent_gene$dfer),]


```
# References